File: android/src/main/java/dragon3/common/DataList.java
package dragon3.common;


import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import dragon3.data.Data;
import mine.MineException;
import mine.io.JsonIO;

/**
 * @author k-saito
 */
public class DataList<T extends Data> {

	private List<T> list;

	private Map<String, T> map;

	@SuppressWarnings("unchecked")
	public DataList(String baseDir, List<String> fileList, Class<?> clazz) {
		super();

		list = new ArrayList<T>();
		map = new HashMap<String, T>();
		
		for (String file : fileList) {
			String filePath = baseDir + file;
			try {
				T[] array = (T[])JsonIO.read(filePath, clazz);
				for (T t : array) {
					list.add(t);
					map.put(t.getId(), t);
				}
			} catch (MineException e) {
				throw new RuntimeException("Dataファイルのロードに失敗しました。fileName:[" + filePath + "]", e);
			}
		}
	}

	public List<T> getList() {
		return list;
	}

	public T getData(String id) {
		if (map.containsKey(id)) {
			return map.get(id);
		} else {
			throw new IllegalArgumentException("illegal data id:" + id);
		}
	}

	public Map<String, String> getIdAndName() {
		Map<String, String> idAndName = new LinkedHashMap<String, String>();

		for (int i=0; i<list.size(); i++) {
			Data data = (Data)list.get(i);
			idAndName.put(data.getId(), data.getName());
		}
		return idAndName;
	}
}

File: core/src/main/kotlin/dragon3/common/DataList.kt
package dragon3.common


import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedHashMap

import dragon3.data.Data
import mine.io.JsonManager

/**
 * @author k-saito
 */
class DataList<T : Data>
@SuppressWarnings("unchecked")
constructor(jsonManager: JsonManager, baseDir: String, fileList: List<String>, clazz: Class<Array<T>>) {

    private val list: MutableList<T>

    private val map: MutableMap<String, T>

    init {

        list = ArrayList<T>()
        map = HashMap<String, T>()

        for (file in fileList) {
            val filePath = baseDir + file
            for (t in jsonManager.read(filePath, clazz)) {
                list.add(t)
                map.put(t.id, t)
            }

        }
    }

    fun getList(): List<T> {
        return list
    }

    fun getData(id: String): T {
        if (map.containsKey(id)) {
            return map[id]!!
        } else {
            throw IllegalArgumentException("illegal data id:" + id)
        }
    }

    val idAndName: Map<String, String>
        get() {
            val idAndName = LinkedHashMap<String, String>()

            for (i in list.indices) {
                idAndName.put(list[i].id, list[i].name)
            }
            return idAndName
        }
}
